plugins/wallet_plugin/wallet_manager.cpp

```diff
@@ -0,0 +1,164 @@
+   std::string password = gen_password();
+
+   auto wallet_filename = dir / (name + file_ext);
+   if (fc::exists( dir / wallet_filename)) {
+      FC_THROW("Wallet with name: ${n} already exists.", ("n", name));
+   }
+
+   wallet_data d;
+   auto wallet = make_unique<wallet_api>(d);
+   wallet->set_password(password);
+   wallet->set_wallet_filename(wallet_filename.string());
+   wallet->unlock(password);
+   wallet->save_wallet_file();
+   wallets.emplace(name, std::move(wallet));
+
+   return password;
+}
+
+void wallet_manager::open(const std::string& name) {
+   check_timeout();
+   wallet_data d;
+   auto wallet = std::make_unique<wallet_api>(d);
+   auto wallet_filename = dir / (name + file_ext);
+   wallet->set_wallet_filename(wallet_filename.string());
+   if (!wallet->load_wallet_file()) {
+      FC_THROW("Unable to open file: ${f}", ("f", wallet_filename.string()));
+   }
+   wallets.emplace(name, std::move(wallet));
+}
+
+std::vector<std::string> wallet_manager::list_wallets() {
+   check_timeout();
+   std::vector<std::string> result;
+   for (const auto& i : wallets) {
+      if (i.second->is_locked()) {
+         result.emplace_back(i.first);
+      } else {
+         result.emplace_back(i.first + " *");
+      }
+   }
+   return result;
+}
+
+std::vector<std::string> wallet_manager::list_keys() {
+   check_timeout();
+   std::vector<std::string> result;
+   for (const auto& i : wallets) {
+      if (!i.second->is_locked()) {
+         result.emplace_back(i.first);
+         const auto& keys = i.second->list_keys();
+         for (const auto& i : keys) {
+            result.emplace_back(i.second);
+         }
+      }
+   }
+   return result;
+}
+
+void wallet_manager::lock_all() {
+   // no call to check_timeout since we are locking all anyway
+   for (auto& i : wallets) {
+      if (!i.second->is_locked()) {
+         i.second->lock();
+      }
+   }
+}
+
+void wallet_manager::lock(const std::string& name) {
+   check_timeout();
+   if (wallets.count(name) == 0) {
+      FC_THROW("Wallet not found: ${w}", ("w", name));
+   }
+   auto& w = wallets.at(name);
+   if (w->is_locked()) {
+      return;
+   }
+   w->lock();
+}
+
+void wallet_manager::unlock(const std::string& name, const std::string& password) {
+   check_timeout();
+   if (wallets.count(name) == 0) {
+      FC_THROW("Wallet not found: ${w}", ("w", name));
+   }
+   auto& w = wallets.at(name);
+   if (!w->is_locked()) {
+      return;
+   }
+   w->unlock(password);
+}
+
+void wallet_manager::import_key(const std::string& name, const std::string& wif_key) {
+   check_timeout();
+   if (wallets.count(name) == 0) {
+      FC_THROW("Wallet not found: ${w}", ("w", name));
+   }
+   auto& w = wallets.at(name);
+   if (w->is_locked()) {
+      FC_THROW("Wallet is locked: ${w}", ("w", name));
+   }
+   w->import_key(wif_key);
+}
+
+chain::SignedTransaction
+wallet_manager::sign_transaction(const chain::SignedTransaction& txn, const chain::chain_id_type& id) {
```

Comment:
It appears that this function is just blindly signing the transaction with all keys in all unlocked wallets? A transaction is only valid if signed with no irrelevant keys, so this is inadequate. The `AuthorityChecker` class can tell you exactly which keys you need to sign with so you sign the transaction correctly.

Another question is: should we be grabbing keys from all wallets? Or should the user specify a particular wallet to get keys from?

Reply:
This was before you pointed me to the `AuthorityChecker` and Dan said to add an api to EOSD to provide the keys.  I'm thinking that interface on `chain_controller` will look like this:

```C++
         /**
          * Determine which public keys are needed to sign the given transaction.
          * @param trx Transaction that requires signature
          * @param candidateKeys Set of public keys to examine for applicability
          * @return Subset of candidateKeys whose private keys should be used to sign transaction
          * @throws fc::exception if candidateKeys does not contain all required keys
          */
         flat_set<public_key_type> get_required_keys( const SignedTransaction& trx, const flat_set<public_key_type>& candidateKeys)const;
```
eosc will get keys from all **unlocked** wallets, call above, then ask walletd to sign with the subset of keys.

My thought is all unlocked wallets. But could add a command for which wallets. thoughts?

Reply:
> eosc will get keys from all unlocked wallets, call above, then ask walletd to sign with the subset of keys.

Makes sense.

What use case is the multiple wallet support aimed at?

Reply:
> What use case is the multiple wallet support aimed at?

Good question. No idea. Just doing as I'm told ;)

Reply:
Who is telling you to support multiple wallets, what is your definition of a wallet?

Reply:
> Who is telling you to support multiple wallets, what is your definition of a wallet?
You did. But apparently, I misunderstood.  I understand a wallet to be a named file of encrypted private keys.

I assumed you wanted multiple wallets so a user could segment his keys and only unlock a particular wallet at a time.

The interface allows a user to create a named wallet. Open a named wallet. Unlock a particular wallet, etc. If there is only one wallet, a name would not be needed.

Here is the comment from the old eosc I was using as a guide:
* eosc unlock walletname PASSWORD
* eosc create wallet walletname PASS1 PASS2
* eosc unlock walletname PASSWORD
* eosc wallets -> prints list of wallets with * next to currently unlocked
* eosc keys -> prints list of private keys
* eosc importkey privatekey -> loads keys
* eosc accounts -> prints list of accounts that reference key
* eosc lock